plugins/linuxplugin/nsplugin/namespaces.go

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nsplugin
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"runtime"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/ligato/cn-infra/logging"
+	intf "github.com/ligato/vpp-agent/plugins/linuxplugin/common/model/interfaces"
+	"github.com/ligato/vpp-agent/plugins/linuxplugin/common/model/l3"
+	"github.com/vishvananda/netns"
+)
+
+// Namespace-related constants
+const (
+	// Configuration namespace for veth interfaces
+	configNamespace = "veth-cfg-ns"
+	// Namespace mount directory
+	netNsMountDir = "/var/run/netns"
+	// Namespace types
+	PidRefNs          = 0
+	MicroserviceRefNs = 1
+	NamedNs           = 2
+	FileRefNs         = 3
+)
+
+// Namespace is a generic representation of typed namespace (interface, arp, etc...)
+type Namespace struct {
+	Type         int32
+	Pid          uint32
+	Microservice string
+	Name         string
+	FilePath     string
+}
+
+// NamespaceMgmtCtx represents context of an ongoing management of Linux namespaces.
+// The same context should not be used concurrently.
+type NamespaceMgmtCtx struct {
+	lockedOsThread bool
+}
+
+// NewNamespaceMgmtCtx creates and returns a new context for management of Linux namespaces.
+func NewNamespaceMgmtCtx() *NamespaceMgmtCtx {
+	return &NamespaceMgmtCtx{lockedOsThread: false}
+}
+
+// CompareNamespaces is a comparison function for "Namespace" type.
+func (ns *Namespace) CompareNamespaces(nsToCompare *Namespace) int {
+	if ns == nil || nsToCompare == nil {
+		if ns == nsToCompare {
+			return 0
+		}
+		return -1
+	}
+	if ns.Type != nsToCompare.Type {
+		return int(ns.Type) - int(nsToCompare.Type)
+	}
+	switch ns.Type {
+	case PidRefNs:
+		return int(ns.Pid) - int(ns.Pid)
+	case MicroserviceRefNs:
+		return strings.Compare(ns.Microservice, nsToCompare.Microservice)
+	case NamedNs:
+		return strings.Compare(ns.Name, nsToCompare.Name)
+	case FileRefNs:
+		return strings.Compare(ns.FilePath, nsToCompare.FilePath)
+	}
+	return 0
+}
+
+// GenericToIfNs converts generic namespace to interface-type namespace
+func (ns *Namespace) GenericToIfNs() (*intf.LinuxInterfaces_Interface_Namespace, error) {
```

Comment:
Rename this to `GenericToIfaceNs()` for clarity, otherwise might be confused with if as condition.